### PR TITLE
feat(eslint-plugin): [prefer-readonly] private fields support

### LINKED ISSUE #7126: Enhancement: [prefer-readonly] check #private class fields
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&showAST=es&code=MYGwhgzhAEAqCmEAu0DeBYAUNaAHATgJYBuYS80A+gB7QC80AjANxY4DEAnvU65m9ADm8FNQAUASjTR8IgK74AdtCQALQhAB0NZtAC+A4Sk6TpspAuVqNmrroOY9QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIXRRe0EdEjgwAXxBKgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
class Test {
  private _x = 1;
  #y = 1;

  get x() { return this._x; }
  get y() { return this.#y; }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/prefer-readonly": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Triggers on `private field` and `#field` class fields.

### Actual Result

Triggers on `private field` but not on `#field` class fields.

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by Jamesernator**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7126
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This adds private field support to `prefer-readonly` with the same behaviour as for classic TS private properties.

The tests are just duplicates of the existing tests, but with `private name` replaced with `#name`.

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/prefer-readonly.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ class Container {
   // These member variables could be marked as readonly
   private neverModifiedMember = true;
   private onlyModifiedInConstructor: number;
+  #neverModifiedPrivateField = 3;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/prefer-readonly.md:25]: [Docs] Earlier in the page, it says:

> Member variables with the privacy `private` are never permitted to be modified outside of their declaring class. If that class never modifies their value, they may safely be marked as `readonly`.
>
> This rule reports on private members are marked as `readonly` if they're never modified outside of the constructor.

Can you add a mention of `#` privates somewhere in there too? The explanation is no longer accurate now that they're also considered by the rule.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
> +5/-1 in implementation
> +907 in tests

Love it 😂. Super thorough tests - nicely done! 👏 

Just requesting changes on a bit of docs too please. Thanks!

**Comment by JoshuaKGoldberg** [comment]
👋 Hey @Jamesernator! Just checking in, is this still something you have time for? No worries if not - I just don't want to leave it hanging.

**Comment by Jamesernator** [comment]
> 👋 Hey @Jamesernator! Just checking in, is this still something you have time for? No worries if not - I just don't want to leave it hanging.

I've added a small docs change to mention it now affects both typescript `private` modifiers and private fields.

File: packages/eslint-plugin/docs/rules/prefer-readonly.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ description: "Require private members to be marked as `readonly` if they're neve
 >
 > See **https://typescript-eslint.io/rules/prefer-readonly** for documentation.
 
-Member variables with the privacy `private` are never permitted to be modified outside of their declaring class.
+Private member variables (whether using the `private` modifier, or private fields) are never permitted to be modified outside of their declaring class.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/prefer-readonly.md:9]: [Docs] Just a bit of nitpicking: the `,` is unnecessary I think, and for consistency + clarity it'd be good to show that `#` is the private fields thing.

```suggestion
Private member variables (whether using the `private` modifier or private `#` fields) are never permitted to be modified outside of their declaring class.
```

**Review by JoshuaKGoldberg — APPROVED** [review]
Awesome, thanks! The tests are just as awesomely thorough as before. 🚀

**Pull request merged**